### ISSUE TITLE
False-positive? CA5379: Do not use weak key derivation function algorithm

### ISSUE THREAD

**Issue by mikernet** [dotnet/roslyn-analyzers#4110]
### Analyzer

**Diagnostic ID**: [CA5379](https://docs.microsoft.com/en-us/visualstudio/code-quality/ca5379): `Do not use weak key derivation function algorithm`

### Analyzer source

**NuGet Package**: [Microsoft.CodeAnalysis.FxCopAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers)

**Version**: v3.3.0 (Latest)

### Describe the bug

CA5379 is triggered even though neither SHA1 nor MD5 is used anywhere. The algorithm is passed in via a property, like so:

```c#
internal class Pbkdf2PasswordHashAlgorithm : PasswordHashAlgorithm
{
    public HashAlgorithmName AlgorithmName { get; }

    public int HashSize { get; }

    public Pbkdf2PasswordHashAlgorithm(string algorithmId, HashAlgorithmName algorithmName, int hashSize) : base(algorithmId)
    {
        AlgorithmName = algorithmName;
        HashSize = hashSize;
    }

    public override byte[] CreateHash(byte[] password, byte[] salt, int iterations)
    {
        // CA5379 on this line:
        using var rfc2898 = new Rfc2898DeriveBytes(password, salt, iterations, AlgorithmName);
        return rfc2898.GetBytes(HashSize);
    }
```

### Expected behavior

Do not trigger a warning on the marked line.

### Actual behavior

Triggers the specified warning.

## Additional context

If this is by design (which I would find odd) then the rule documentation should at least be updated to specify that if the algorithm is not known at compile-time then the rule is also triggered. That said, I don't think this should trigger the rule...rather it should be triggered anywhere `HashAlgorithmName.SHA1` or `HashAlgorithmName.MD5` is referenced, or if a different constructor for `Rfc2898DeriveBytes` is used where the algorithm is implicitly selected to be `SHA1`, which is not the case here.

**Comment by mavasani** [comment]
Tagging @dotpaul 

Personally, I think this is by design as security rules are generally more aggressive - fire whenever unsure that safe APIs are used. Users should add a `Debug.Assert` or a source suppression in the invoking method to explicitly annotate/document their source code that they are using a safe API.

**Comment by mikernet** [comment]
What do you mean by `Debug.Assert`? This does not remove the warning so I'm guessing you mean something else:

```c#
public override byte[] CreateHash(byte[] password, byte[] salt, int iterations)
{
    var a = AlgorithmName;
    Debug.Assert(a != HashAlgorithmName.MD5 && a != HashAlgorithmName.SHA1);

    using var rfc2898 = new Rfc2898DeriveBytes(password, salt, iterations, a); // still warns here.
    return rfc2898.GetBytes(HashSize);
}
```

**Comment by dotpaul** [comment]
I'll update the docs to note its behavior and rename the rule to "Ensure key derivation function hash algorithm is strong" or something. It's always a judgment call as to what an analyzer should do when it can't definitively determine the state. For this particular rule, I don't think Rfc2898DeriveBytes appears in code frequently (maybe once or a few times per code base), so false positives are more acceptable.

**Comment by joemey** [comment]
Another problem with `Rfc2898DeriveBytes` is that the constructor overloads which accepts a `HashAlgorithmName` have been added with `.Net Standard 2.1` and `.NET 4.7.2` and are not available on prior versions.

CA5379 will be generated when compiling for a version where it is not possible to specifiy the `HashAlgorithmName`
